### DISCUSSION TITLE
How to get the field name of a node

### DISCUSSION THREAD

**Discussion by meain** [emacs-tree-sitter/elisp-tree-sitter#173] [Q&A]
I was wondering if there was a way to get the field name of a node. I am not sure if I using the correct terminology, but I was wondering if I can give the `type_identifier` node and get something like `:type` or `"type"` out of it? I am aware of `tsc-get-child-by-field`, but I wanted to lookup the field name if any for a specific node. I was not able to find anything in the elisp files or follow this function.

Something with the signature `Node -> Optional<Field>`

![Screenshot 2021-09-12 at 6 57 21 PM](https://user-images.githubusercontent.com/14259816/132989423-4d6a518c-b04e-47ca-bd30-a092225140e3.png)

**Comment by ubolonton** [comment]
There is currently no API to get a node's field name under its parent node. As a workaround, you can use this implementation (it's a bit convoluted, and not as efficient as a Rust implementation could be):

```emacs-lisp
(defun tsc-node-field (node)
  (when-let* ((parent (tsc-get-parent node))
              (cursor (tsc-make-cursor parent))
              (_  (tsc-goto-first-child cursor))
              (child (tsc-current-node cursor)))
    (while (and child (not (tsc-node-eq child node)))
      (setq child (if (tsc-goto-next-sibling cursor)
                      (tsc-current-node cursor)
                    nil)))
    (tsc-current-field cursor)))
```

What is your specific use case, though?

**Reply by meain** [reply]
To be frank, I was just playing around with the debug-mode and trying to add some more info onto it. I was thinking about a info panel via popups or in modeline which would have this when I am on a node in the code.

Btw, is the rust interaction with the elisp part documented anywhere. Where should I be looking into if I were to add something like this into the rust code?

**Reply by ubolonton** [reply]
> To be frank, I was just playing around with the debug-mode and trying to add some more info onto it. I was thinking about a info panel via popups or in modeline which would have this when I am on a node in the code.

That sounds pretty useful!

> Btw, is the rust interaction with the elisp part documented anywhere. Where should I be looking into if I were to add something like this into the rust code?

Basically you would write a new Rust function and mark it with `#[defun]` so that it's callable from Lisp. The Rust code for node APIs are in [core/src/node.rs](https://github.com/emacs-tree-sitter/elisp-tree-sitter/blob/0.15.2/core/src/node.rs#L226).  You can check out [`emacs-module-rs` doc](https://ubolonton.github.io/emacs-module-rs/latest/functions.html) for more details.
